Allow binding factory functions as constructors

jagerman opened the pull request:
Edit:

Redid this (a few times).  The interface is now:

```C++
    cls.def(py::init(&factory_function));
```

where `factory_function` is some pointer- or holder-generating factory function or lambda of the type that `cls` binds, such as:

```C++
    cls.def(py::init([](int a) { return new MyClass(a); }));
```

Internally, this still results in a (non-static) `__init__` method, but we handle the C++-static-factory <-> python method translation by calling the factory function, then stealing the resulting object internal pointer and holder.

jagerman commented:
(Credit to @YannickJadoul and @bmerry for the idea).

jagerman commented:
Also worth noting: this isn't affected by #804, because the `def_static("__init__", ...)` support here doesn't actually bind a *python*-side static method, so `def("__init__")` and `def_static("__init__")` can be mixed.

wjakob commented:
This strikes me a fairly intrusive change to support a relatively narrow set of situations. (In particular, the ``instance_swap()`` function, and that we need to stick something into the dispatch code regardless of whether this feature is used or not.)

As an alternative, could we have something like:

```cpp
m.def("__init__", py::factory_adapter(&Example::create));
```

where the ``factory_adapter`` either uses an in-place move constructor, or it takes a ``py::object`` as ``self`` argument and then does something similar to ``instance_swap``.

jagerman commented:
> As an alternative, could we have something like:
>
> ```C++
> m.def("__init__", py::factory_adapter(&Example::create));
> ```
> where the factory_adapter either uses an in-place move constructor, or it takes a py::object as self argument and then does something similar to instance_swap.

Yeah, that should be doable.  I'll take another shot at this in the next day or two.

jagerman commented:
And that MSVC job failure is more bizarre than ever.

YannickJadoul commented:
>it takes a py::object as self argument and then does something similar to instance_swap.

I don't want to come across whiny, but could I bring up that gist I've mentioned in the gitter chat, again?
https://gist.github.com/YannickJadoul/38dc4ef8f0faa56fa10824626886edab

I'm sure it could be improved a lot with better knowledge of pybind11, but what it more or less does is:
- Define a class `unconstructed_holder` which takes care of the deallocation of the allocated memory for the instance, then constructs the holder immediately.
- Define `type_caster<unconstructed_holder>` that is only able to load a `py::instance`, not just extracting the pointer to the instance but also the memory reserved for the holder

jagerman commented:
Redesigned this.

The redesign is far less intrusive, and avoids some unnecessary Python object creation in several cases.

I still want to add some tests to verify all the various edge cases (e.g. for things like returning an alias pointer or value, or a subclass pointer).

wjakob commented on docs/advanced/classes.rst:
```diff
@@ -366,6 +366,35 @@ In other words, :func:`init` creates an anonymous function that invokes an
+        // Bind an existing pointer-returning factory function:
+        .def(py::init_factory(&Example::create))
+        // Similar, but returns the pointer wrapped in a holder:
+        .def("__init__", []() {
```
``init_factory`` missing.

wjakob commented:
Nifty -- +1 for covering all the different kinds of things that could be returned. Since the ``detail::factory_init`` got quite large, I'm wondering if it might make sense to move it into a separate header file (e.g. ``pybind11/factory.h``) since it addresses a fairly specific requirement.

jagerman commented:
As far as organization, moving it to another header is a good idea.  It can't *quite* be all moved, though: the `class_::init(py::init_factory &&)` member function is still needed (because it needs to pass the `class_<...>` type to `execute()`).

wjakob commented:
Yep, that is to be expected. With a forward declaration, it should be possible though, no?.

jagerman commented:
I assume you meant making it an optional (i.e. requires an extra include before using) header.

YannickJadoul commented:
I don't know about all the subtleties of handling e.g. `py::handle` or `py::object`, but this seems a really nice solution to the problem to me, especially the way it hardly requires an internal change :-)
Trying this out on my code and test case I'd written works like a charm.

I've got two possibly useful things:
- As far as I know, `init_holder` is still called afterwards, which might mess things up by overwriting the holder once again? I.e., since the `owned` field of the instance is still true, `init_holder` will think it needs to wrap the pointer again in a holder? (But maybe I've overlooked something.)
- Just a detail, but nevertheless: I believe your `alias_constructible` does not need SFINAE but can do with just a `using alias_constructible = std::integral_constant<bool, that_long_boolean_expression>` ?

But for the rest, just "wow", again!

jagerman commented:
Re: `init_holder`: I added a check to the if-constructor-then-init-holder call in `dispatch`, but apparently it got lost when I was cleaning up the commit: good catch!

The SFINAE was necessary because the `bool_constant` version fails when the type has no alias because `Alias<Class>` is `void` and so `Alias<Class> &&` is taking a reference to void, which is invalid.

YannickJadoul commented:
>The SFINAE was necessary because the bool_constant version fails when the type has no alias because Alias<Class> is void and so Alias<Class> && is taking a reference to void, which is invalid.

Aha, my bad, sorry

But thanks again for the quick and elegant solution!

jagerman commented:
I've added various tests and fixes related to value and alias initialization, and moved the tests out into a new `test_factory_constructors` rather than bulking up `method_and_attributes` with things that are only sort-of methods.

wjakob commented:
This probably needs to be rebased on the latest master, it looks like the PR has some conflicts and contains MI-related changes.

jagerman commented:
Fixed.  (This was already rebased against #812--I just had to delete all the duplicate commits).

jagerman commented:
(There's still an outstanding MSVC issue I need to track down here).

jagerman commented:
Oh, it was just an annoying warning (because `factory.h` is outside `pybind11.h`'s ignore warnings stack).

dean0x7d commented on include/pybind11/factory.h:
```diff
@@ -0,0 +1,269 @@
+
+    template <typename Class, typename SFINAE = void> struct alias_constructible : std::false_type {};
+    template <typename Class> struct alias_constructible<Class, enable_if_t<Class::has_alias &&
+            std::is_convertible<Return, Alias<Class>>::value && std::is_constructible<Alias<Class>, Alias<Class> &&>::value>>
```
`std::is_constructible<Alias<Class>, Alias<Class> &&>` == `std::is_move_constructible<Alias<Class>>` which might better reveal the intent.

jagerman replied:
Good point.  I had thought `is_move_constructible` was more specifically "has move constructor", but I see that's not the case.

dean0x7d commented on include/pybind11/factory.h:
```diff
@@ -0,0 +1,269 @@
+        : std::true_type {};
+    template <typename Class> using cpp_constructible = bool_constant<
+        !(Class::has_alias && std::is_base_of<Alias<Class>, Return>::value) &&
+        std::is_convertible<Return, Cpp<Class>>::value && std::is_constructible<Cpp<Class>, Cpp<Class> &&>::value>;
```
Same here.

jagerman commented:
> Did I overlook something?

The idea I had in mind for that was making it easier to return values that come from Python.  So, for example, a particular constructor could defer to a Python factory function to create the instance without having to do the unboxing in a C++ factory function, while wrapping the whole thing up by exposing it as a constructor.  I'll add a couple of example of this to the tests.

I do agree that making a pure C++ factory function wrap the return in a `py::object` seems fairly pointless.

jagerman commented:
One last question: should `factory.h` get included by default by `pybind11.h`?  It doesn't actually add any dependencies, and is all templated code, so shouldn't really affect anything if it isn't actually being used.

jagerman commented:
> > Did I overlook something?

Looking at the pypy failure, I think you're right that this isn't useful; there's no guarantee a python factory function will be uniquely referenced (though seems to be in cpython).  I'll go ahead and remove it.

dean0x7d commented:
> @jagerman
> One last question: should factory.h get included by default by pybind11.h? It doesn't actually add any dependencies, and is all templated code, so shouldn't really affect anything if it isn't actually being used.

I think @wjakob's suggestion for moving out the header was for compilation performance (even just parsing C++ is difficult because so much is context sensitive). I think that if some of the metafunctions could be simplified a bit (more below) it might be worthwhile to include it in `pybind11.h` just to better organize the features (make them easier to teach). Specifically, with the separate header, there are 3 ways of creating constructors which look like this:
```c++
.def(py::init<Args...>())
.def("__init__", [](Class&, Args...) { ... })
.def(py::init_factory([](Args...) { ... }))
```
where the last one is a superset of the second one, but requires an extra header. The 3 methods also use wildly different syntax. If the factory is made into a first-class feature, this could become more uniform:
```c++
.def(py::init<Args...>())
.def(py::init([](Args...) { ... }))
```
Note that `init_factory` is renamed to `init` (just an overload of the existing `init` function). The factory `init` is a type-safe alternative to placement `new`, so it would be preferable to teach this in the docs (leave the raw `__init__` as an expert-level feature).

To make the factory implementation a bit more light-weight, would it be possible to rely more on simple overload resolution than the metafunctions + tags? To me, it looks like `construct` overloads are already unique, so the `factory_type` metafunction may not be completely necessary.

YannickJadoul commented:
> The factory `init` is a type-safe alternative to placement `new`

Just as a side note: this type-safety only relates to the C++ side of things. Because of the `py::handle` as first argument and the implicit assumption that this will always be of the right type, you cán actually break this at 'Python runtime' by doing something silly like `X.__init__(Y.__new__(Y))`.
(I know, very far fetched, but pybind11 won't stop you from segfaulting your Python instance, while - I think - the standard `init` would.)

dean0x7d commented:
> Because of the py::handle as first argument and the implicit assumption that this will always be of the right type, you cán actually break this at 'Python runtime' by doing something silly like `X.__init__(Y.__new__(Y))`.

Interesting point. I guess a quick type check should be able to prevent that.

jagerman commented:
I took another stab at this in light of @dean0x7d 's comments and some other issues to simplify the interface and implementation.  Most of the issues related to alias classes.  The implementation is a bit lighter, but the code logic grew a bit to handle the special cases aliases bring up.

Summary of changes:
- renamed `py::init_factory` to `py::init`.  This is easy to revert, but this doesn't conflict with the templated no-argument `py::init<...>()`, while keeping the naming nicer (especially if we decide to roll this into the core, documented approach replacing raw placement-new `__init__`s).
- factories can no longer return types *convertible* to the base/alias class: they can now return only a {base,alias} {value,pointer}, a holder, or a shared_ptr<T> to some base or derived `T` (only applies when the class has a `shared_ptr` holder to begin with).
- cutting out the convertible support allowed the implementation to be a bit lighter by avoiding the need for the tag classes (which were needed to prioritize ambiguous calls).
- If returning a base class (whether by value or pointer) when an alias is required (i.e. python-side inheritance and a declared alias), a dynamic_cast to the alias is attempted (for the pointer version); if it fails, or if returned by value, an Alias(Class &&) constructor needs to be defined (otherwise a type_error is thrown).
- for holder returns when an alias is required, we try a dynamic_cast of the pointer to the alias; if it fails, we raise an error.
- a new `py::init(class_factory, alias_factory)` function is available that takes two factories: the first is called when an alias is not needed, the second when it is.  (This, in particular, is how holder-returning factories can still be used, but puts the onus on the caller to sort out the base/alias creation logic).

(I'm holding off on the writing the documentation until the implementation looks good).

bmerry commented:
I'm not clear why one would want to accept a factory function that returns a `shared_ptr<T>` where T is a base class of the class being defined (as opposed to a derived class). That sounds like it discards type safety, and if the user really wants to do that, they can write it in a lambda that does the pointer cast themselves.

However, I have barely used pybind11 and I don't really understand the aliases concept, so I might be missing something.

jagerman commented:
@bmerry - The story is the same for both base class pointers and `std::shared_ptr<Base>`: it doesn't discard type safety as long as the type is polymorphic.  It *does* need a `dynamic_cast` (via `std::dynamic_pointer_cast` for a `shared_ptr`) to do a runtime RTTI check as to whether the pointer actually is the derived type, and the get the derived pointer (which doesn't have to be equal to the base pointer, particularly when multiple inheritance is involved).

bmerry commented:
I was thinking more of static than dynamic type safety. If I saw code that used a factory function returning `shared_ptr<Base>` being used as a constructor for the Python wrapper of Derived, my first guess would be that either
(a) The factory only ever returns `shared_ptr<Derived>`, and ought to have been declared as such; or
(b) The factory doesn't always return `shared_ptr<Derived>`, and the user either ought to be wrapping Base rather than Derived, or they meant to use a different factory function.

Maybe there are legitimate use cases where one wishes to do this, and get an exception thrown if the Python caller passes a combination of arguments that causes the factory not to spit out a `shared_ptr<Derived>`, but my (prejudiced, non-scientific) feeling is that this would be much rarer than the cases above. Removing the support for it from pybind11 would help catch errors at compile time. It wouldn't prevent users implementing this use case, because the user can write a wrapper around his/her factory function that does the dynamic cast, and perhaps raises a more application-aware error whose message explains why the constructor is failing.

bmerry commented on docs/advanced/classes.rst:
```diff
@@ -366,6 +366,51 @@ In other words, :func:`init` creates an anonymous function that invokes an
+        .def(py::init_factory(&Example::create))
+        // Similar, but returns the pointer wrapped in a holder:
+        .def(py::init_factory([](std::string arg) {
+            return std::unique_ptr<Example>(new Example(arg, "another arg"));
```
Should this be changed init_factory -> init now?

YannickJadoul replied:
Yes, but @jagerman said he'd wait before updating that
> (I'm holding off on the writing the documentation until the implementation looks good).

bmerry commented on include/pybind11/factory.h:
```diff
@@ -0,0 +1,316 @@
+    CReturn (*)(CArgs...), AReturn (*)(AArgs...));
+
+template <typename... Func> using init_factory_t = decltype(init_factory_decltype<Func...>(
+    (typename detail::remove_class<decltype(&std::remove_reference<Func>::type::operator())>::type *) nullptr...));
```
When I tried this out with a plain function it didn't work. I suspect it's because a plain function doesn't have a `::operator()` when it tries to expand this. I can't think of a simple way to get the return type of either a function or a function object (`result_of` requires you to know the calling argument types). I guess it'll just need another traits helper class to handle the various cases.

bmerry replied:
Your fix for this is working for me - thanks.

dean0x7d commented on include/pybind11/factory.h:
```diff
@@ -0,0 +1,322 @@
+
+NAMESPACE_BEGIN(detail)
+
+inline static void init_factory_no_nullptr(void *ptr) {
```
Shouldn't be `static`.

dean0x7d commented:
> @bmerry said:
> I'm not clear why one would want to accept a factory function that returns a shared_ptr<T> where T is a base class of the class being defined (as opposed to a derived class).
>
> If I saw code that used a factory function returning shared_ptr<Base> being used as a constructor for the Python wrapper of Derived, my first guess would be that either
(a) The factory only ever returns shared_ptr<Derived>, and ought to have been declared as such; or
(b) The factory doesn't always return shared_ptr<Derived>, and the user either ought to be wrapping Base rather than Derived, or they meant to use a different factory function.

I think this is a good point. `init_factory` is checking everything at runtime, but really, the user should already know this when binding a constructor. A function that can return an invalid type depending on the given arguments isn't a good choice for a constructor anyway.

A pretty large portion of the code is dedicated to supporting downcasting (+ double that because of the alias constructor). Removing it would be major simplification.

jagerman commented:
I've updated this PR to slim this down some more by removing the downcasting support (except for the `Base *` -> `Alias *` cast attempt when we need an Alias), which simplifies it quite a bit.

Thoughts, now, on including this by default (and updating to docs to prefer this over placement-new in simpler cases)?

Any other feedback?

jagerman commented:
Oops, I missed some of @dean0x7d's comments, I'll update with them shortly.

jagerman commented on include/pybind11/common.h:
```diff
@@ -509,6 +509,31 @@ template <typename T>
+    std::is_pointer<T>::value && std::is_function<typename std::remove_pointer<T>::type>::value>;
+
+// Extracts the function signature from a function, function pointer or lambda.
+template <typename F, typename SFINAE = void> struct function_signature {};
```
This fails to compile because of the `&F::operator()` when `F` is a plain function pointer; any idea for a workaround?

dean0x7d replied:
Untested, but try:
```c++
typename std::conditional_t<
    std::is_pointer<F>::value || std::is_member_pointer<F>::value,
    remove_pointer<F>, // Note: no `_t`
    strip_function_object<F>
>::type // deferred `::type` access
```
where:
```c++
template <typename F>
struct strip_function_object {
    using type = typename remove_class<decltype(&F::operator())>::type;
}
```

jagerman replied:
That worked.  I actually dropped the whole `struct function_signature` and just declared `using function_signature_t = conditional_t<...>`.

jagerman commented:
I've rebased and updated this PR with the new testing style.

As for whether or not to include it by default, I ran some benchmarks building the test suite, testing both compiler CPU time and peak memory usage (RSS).  The only change between the two runs below is whether `#include <pybind11/factory.h>` is added to `pybind11_tests.h` (i.e. included in every compilation unit).

Here are the gcc 7.1.0 results:
```
                                   gcc-wo-factory.txt [n=50]                     gcc-w-factory.txt [n=50]              
                                ================================   ====================================================
                                 CPU   (s.d. )   MaxRSS ( s.d. )    CPU   (s.d. ) (   ±   )   MaxRSS ( s.d. ) (   ±   )
                                --------------   ---------------   ------------------------   -------------------------
                   (linking):   27.865 (0.119)   288764 (  82.8)   27.866 (0.126) ( +0.01%)   288786 (  67.6) ( +0.01%)
              pybind11_tests:    2.201 (0.029)   294996 ( 132.3)    2.206 (0.029) ( +0.22%)   295576 (  51.9) ( +0.20%)
                test_buffers:    2.557 (0.025)   310068 (  79.9)    2.567 (0.027) ( +0.36%)   310264 ( 118.4) ( +0.06%)
        test_builtin_casters:    2.969 (0.038)   340145 (  98.6)    2.967 (0.034) ( -0.07%)   339710 (  78.9) ( -0.13%)
          test_call_policies:    1.882 (0.031)   254697 (  43.4)    1.882 (0.026) ( -0.01%)   255205 (  77.1) ( +0.20%)
              test_callbacks:    2.735 (0.033)   310796 (  50.5)    2.740 (0.030) ( +0.18%)   310648 ( 131.2) ( -0.05%)
                 test_chrono:    1.762 (0.027)   252543 ( 125.8)    1.771 (0.027) ( +0.53%)   252720 (  70.0) ( +0.07%)
                  test_class:    3.057 (0.035)   356680 (  54.0)    3.069 (0.039) ( +0.38%)   356962 (  54.4) ( +0.08%)
test_constants_and_functions:    2.136 (0.028)   291964 ( 100.5)    2.133 (0.029) ( -0.18%)   292425 ( 107.2) ( +0.16%)
              test_copy_move:    3.064 (0.035)   342694 ( 134.9)    3.061 (0.034) ( -0.10%)   342977 (  60.1) ( +0.08%)
      test_docstring_options:    1.761 (0.030)   242729 ( 134.8)    1.755 (0.030) ( -0.32%)   243192 ( 140.1) ( +0.19%)
                  test_eigen:    9.299 (0.091)   728918 (  57.2)    9.283 (0.075) ( -0.17%)   729366 (  58.9) ( +0.06%)
                   test_enum:    3.316 (0.035)   354021 (  92.6)    3.312 (0.037) ( -0.10%)   353723 (  75.8) ( -0.08%)
                   test_eval:    1.524 (0.021)   215105 (  60.4)    1.527 (0.023) ( +0.20%)   215526 (  54.1) ( +0.20%)
             test_exceptions:    1.459 (0.026)   208468 (  67.0)    1.463 (0.019) ( +0.23%)   208796 (  78.5) ( +0.16%)
   test_factory_constructors:    4.581 (0.050)   418475 (  53.9)    4.564 (0.033) ( -0.38%)   418496 (  60.1) ( +0.01%)
    test_kwargs_and_defaults:    2.121 (0.023)   287457 (  52.0)    2.121 (0.027) ( +0.03%)   287698 (  58.3) ( +0.08%)
 test_methods_and_attributes:    4.541 (0.043)   390432 (  55.8)    4.545 (0.051) ( +0.09%)   390622 (  52.6) ( +0.05%)
                test_modules:    2.200 (0.024)   296003 (  51.6)    2.205 (0.026) ( +0.25%)   296336 (  93.7) ( +0.11%)
   test_multiple_inheritance:    4.896 (0.051)   454484 (  44.4)    4.896 (0.051) ( +0.02%)   454642 (  73.2) ( +0.03%)
            test_numpy_array:    4.222 (0.044)   370925 (  56.9)    4.218 (0.039) ( -0.09%)   371269 (  98.1) ( +0.09%)
           test_numpy_dtypes:    3.072 (0.031)   333738 ( 132.7)    3.067 (0.033) ( -0.14%)   333883 ( 109.4) ( +0.04%)
        test_numpy_vectorize:    3.168 (0.038)   342012 (  82.3)    3.172 (0.037) ( +0.15%)   341343 (  71.4) ( -0.20%)
           test_opaque_types:    2.180 (0.028)   295067 (  84.8)    2.189 (0.033) ( +0.41%)   295338 ( 122.7) ( +0.09%)
   test_operator_overloading:    2.877 (0.030)   334324 ( 135.7)    2.881 (0.043) ( +0.14%)   334708 (  98.7) ( +0.11%)
               test_pickling:    1.853 (0.030)   258136 ( 108.8)    1.855 (0.024) ( +0.12%)   258567 (  76.1) ( +0.17%)
                test_pytypes:    2.088 (0.029)   270000 (  46.6)    2.091 (0.024) ( +0.13%)   270404 ( 105.9) ( +0.15%)
test_sequences_and_iterators:    4.152 (0.037)   377033 (  62.2)    4.147 (0.041) ( -0.12%)   378544 (  79.7) ( +0.40%)
              test_smart_ptr:    4.205 (0.036)   378643 (  69.6)    4.212 (0.037) ( +0.19%)   378940 (  53.0) ( +0.08%)
                    test_stl:    3.546 (0.032)   366140 (  89.6)    3.549 (0.038) ( +0.08%)   366496 ( 111.9) ( +0.10%)
            test_stl_binders:   11.039 (0.090)   806419 ( 145.5)   11.037 (0.088) ( -0.02%)   806921 ( 186.6) ( +0.06%)
      test_virtual_functions:    4.029 (0.032)   377739 (  74.1)    4.039 (0.035) ( +0.25%)   377863 (  63.5) ( +0.03%)
```

The takeaway is that the CPU usage increase is basically indistinguishable from noise (19 increases, 11 decreases, not counting the added factory test); the compiler memory usage increases look slightly less like noise (26 increased, 4 decreased), but the impact is still quite small (only one exceeds 0.2% increase).

For clang 4.0.1 the results are similarly miniscule.  (CPU times increase more consistently positive than gcc, but it is still a very small increase, with pretty much everything under half a percent).
```
                                  clang-wo-factory.txt [n=50]                   clang-w-factory.txt [n=50]             
                                ================================   ====================================================
                                 CPU   (s.d. )   MaxRSS ( s.d. )    CPU   (s.d. ) (  +/-  )   MaxRSS ( s.d. ) (  +/-  )
                                --------------   ---------------   ------------------------   -------------------------
                   (linking):   37.108 (0.616)   250701 (6422.2)   36.970 (0.585) ( -0.37%)   247911 (5011.3) ( -1.11%)
              pybind11_tests:    2.492 (0.032)   149898 ( 373.8)    2.504 (0.028) ( +0.50%)   150165 ( 330.1) ( +0.18%)
                test_buffers:    2.704 (0.034)   159831 ( 247.3)    2.707 (0.031) ( +0.10%)   160006 ( 267.3) ( +0.11%)
        test_builtin_casters:    3.256 (0.033)   182829 ( 217.3)    3.259 (0.034) ( +0.10%)   182961 ( 202.6) ( +0.07%)
          test_call_policies:    2.161 (0.026)   136706 ( 203.1)    2.174 (0.032) ( +0.57%)   136876 ( 192.6) ( +0.12%)
              test_callbacks:    3.058 (0.031)   166637 ( 317.2)    3.068 (0.035) ( +0.31%)   166780 ( 331.1) ( +0.09%)
                 test_chrono:    1.910 (0.021)   138702 ( 221.2)    1.916 (0.025) ( +0.30%)   138812 ( 196.0) ( +0.08%)
                  test_class:    3.400 (0.029)   189369 ( 215.2)    3.397 (0.034) ( -0.10%)   189575 ( 161.1) ( +0.11%)
test_constants_and_functions:    2.433 (0.021)   149718 ( 254.9)    2.441 (0.030) ( +0.34%)   149864 ( 251.1) ( +0.10%)
              test_copy_move:    3.289 (0.032)   189856 ( 200.6)    3.290 (0.036) ( +0.05%)   190092 ( 170.7) ( +0.12%)
      test_docstring_options:    2.018 (0.025)   133543 ( 238.5)    2.022 (0.025) ( +0.16%)   133715 ( 189.7) ( +0.13%)
                  test_eigen:    9.780 (0.073)   412217 ( 248.5)    9.794 (0.085) ( +0.15%)   412457 ( 304.5) ( +0.06%)
                   test_enum:    3.646 (0.024)   192517 ( 198.4)    3.666 (0.046) ( +0.55%)   192678 ( 188.9) ( +0.08%)
                   test_eval:    1.758 (0.018)   124507 ( 214.8)    1.764 (0.023) ( +0.36%)   124641 ( 174.6) ( +0.11%)
             test_exceptions:    1.648 (0.020)   122190 ( 166.4)    1.653 (0.018) ( +0.32%)   122422 ( 195.0) ( +0.19%)
   test_factory_constructors:    5.396 (0.052)   235282 ( 228.5)    5.393 (0.052) ( -0.06%)   235291 ( 214.6) ( +0.00%)
    test_kwargs_and_defaults:    2.387 (0.024)   150717 ( 383.2)    2.397 (0.024) ( +0.39%)   150876 ( 405.2) ( +0.11%)
 test_methods_and_attributes:    4.864 (0.045)   228137 ( 241.3)    4.861 (0.050) ( -0.05%)   228315 ( 249.5) ( +0.08%)
                test_modules:    2.561 (0.025)   150216 ( 357.0)    2.565 (0.027) ( +0.16%)   150489 ( 398.9) ( +0.18%)
   test_multiple_inheritance:    4.973 (0.051)   259231 ( 359.4)    4.972 (0.039) ( -0.03%)   259470 ( 416.0) ( +0.09%)
            test_numpy_array:    4.740 (0.039)   220536 ( 277.9)    4.731 (0.053) ( -0.19%)   220680 ( 217.6) ( +0.07%)
           test_numpy_dtypes:    3.595 (0.039)   181304 ( 324.2)    3.604 (0.038) ( +0.26%)   181479 ( 250.1) ( +0.10%)
        test_numpy_vectorize:    3.561 (0.034)   188478 ( 183.4)    3.567 (0.040) ( +0.18%)   188645 ( 207.6) ( +0.09%)
           test_opaque_types:    2.493 (0.029)   153162 ( 477.6)    2.501 (0.031) ( +0.32%)   153496 ( 455.0) ( +0.22%)
   test_operator_overloading:    3.062 (0.044)   172689 ( 234.3)    3.068 (0.031) ( +0.18%)   172895 ( 192.2) ( +0.12%)
               test_pickling:    2.116 (0.030)   137702 ( 245.7)    2.123 (0.029) ( +0.30%)   137854 ( 206.4) ( +0.11%)
                test_pytypes:    2.725 (0.029)   143312 ( 259.7)    2.728 (0.028) ( +0.09%)   143504 ( 308.4) ( +0.13%)
test_sequences_and_iterators:    4.408 (0.052)   216225 ( 211.2)    4.406 (0.046) ( -0.04%)   216568 ( 191.3) ( +0.16%)
              test_smart_ptr:    4.387 (0.042)   209329 ( 205.3)    4.390 (0.044) ( +0.06%)   209564 ( 205.9) ( +0.11%)
                    test_stl:    3.873 (0.033)   202854 ( 204.3)    3.880 (0.035) ( +0.17%)   202967 ( 214.6) ( +0.06%)
            test_stl_binders:   11.174 (0.082)   439104 ( 215.3)   11.185 (0.093) ( +0.10%)   439334 ( 239.1) ( +0.05%)
      test_virtual_functions:    4.214 (0.045)   211784 ( 263.2)    4.220 (0.040) ( +0.15%)   211997 ( 301.1) ( +0.10%)
```

So a slight CPU impact (well under half a percent for almost everything), and a smaller memory impact than gcc.

---

If this seems acceptable to everyone, I'll go ahead with adding this as a core include from `pybind11.h`, updating the documentation to prefer this over direct `__init__` initializers (relegating `__init__` to an advanced feature), and update the test suite accordingly.


(I've also committed the tools that generated these with quick instructions in https://github.com/pybind/pybind11_compiler_bench)

--- 

For the curious, here's the comparison between gcc/clang:

```
                                   gcc-wo-factory.txt [n=50]                   clang-wo-factory.txt [n=50]             
                                ================================   ====================================================
                                 CPU   (s.d. )   MaxRSS ( s.d. )    CPU   (s.d. ) (  +/-  )   MaxRSS ( s.d. ) (  +/-  )
                                --------------   ---------------   ------------------------   -------------------------
                   (linking):   27.865 (0.119)   288764 (  82.8)   37.108 (0.616) (+33.17%)   250701 (6422.2) (-13.18%)
              pybind11_tests:    2.201 (0.029)   294996 ( 132.3)    2.492 (0.032) (+13.21%)   149898 ( 373.8) (-49.19%)
                test_buffers:    2.557 (0.025)   310068 (  79.9)    2.704 (0.034) ( +5.74%)   159831 ( 247.3) (-48.45%)
        test_builtin_casters:    2.969 (0.038)   340145 (  98.6)    3.256 (0.033) ( +9.67%)   182829 ( 217.3) (-46.25%)
          test_call_policies:    1.882 (0.031)   254697 (  43.4)    2.161 (0.026) (+14.81%)   136706 ( 203.1) (-46.33%)
              test_callbacks:    2.735 (0.033)   310796 (  50.5)    3.058 (0.031) (+11.83%)   166637 ( 317.2) (-46.38%)
                 test_chrono:    1.762 (0.027)   252543 ( 125.8)    1.910 (0.021) ( +8.41%)   138702 ( 221.2) (-45.08%)
                  test_class:    3.057 (0.035)   356680 (  54.0)    3.400 (0.029) (+11.21%)   189369 ( 215.2) (-46.91%)
test_constants_and_functions:    2.136 (0.028)   291964 ( 100.5)    2.433 (0.021) (+13.89%)   149718 ( 254.9) (-48.72%)
              test_copy_move:    3.064 (0.035)   342694 ( 134.9)    3.289 (0.032) ( +7.32%)   189856 ( 200.6) (-44.60%)
      test_docstring_options:    1.761 (0.030)   242729 ( 134.8)    2.018 (0.025) (+14.64%)   133543 ( 238.5) (-44.98%)
                  test_eigen:    9.299 (0.091)   728918 (  57.2)    9.780 (0.073) ( +5.16%)   412217 ( 248.5) (-43.45%)
                   test_enum:    3.316 (0.035)   354021 (  92.6)    3.646 (0.024) ( +9.95%)   192517 ( 198.4) (-45.62%)
                   test_eval:    1.524 (0.021)   215105 (  60.4)    1.758 (0.018) (+15.38%)   124507 ( 214.8) (-42.12%)
             test_exceptions:    1.459 (0.026)   208468 (  67.0)    1.648 (0.020) (+12.92%)   122190 ( 166.4) (-41.39%)
   test_factory_constructors:    4.581 (0.050)   418475 (  53.9)    5.396 (0.052) (+17.79%)   235282 ( 228.5) (-43.78%)
    test_kwargs_and_defaults:    2.121 (0.023)   287457 (  52.0)    2.387 (0.024) (+12.57%)   150717 ( 383.2) (-47.57%)
 test_methods_and_attributes:    4.541 (0.043)   390432 (  55.8)    4.864 (0.045) ( +7.10%)   228137 ( 241.3) (-41.57%)
                test_modules:    2.200 (0.024)   296003 (  51.6)    2.561 (0.025) (+16.41%)   150216 ( 357.0) (-49.25%)
   test_multiple_inheritance:    4.896 (0.051)   454484 (  44.4)    4.973 (0.051) ( +1.58%)   259231 ( 359.4) (-42.96%)
            test_numpy_array:    4.222 (0.044)   370925 (  56.9)    4.740 (0.039) (+12.26%)   220536 ( 277.9) (-40.54%)
           test_numpy_dtypes:    3.072 (0.031)   333738 ( 132.7)    3.595 (0.039) (+17.03%)   181304 ( 324.2) (-45.67%)
        test_numpy_vectorize:    3.168 (0.038)   342012 (  82.3)    3.561 (0.034) (+12.41%)   188478 ( 183.4) (-44.89%)
           test_opaque_types:    2.180 (0.028)   295067 (  84.8)    2.493 (0.029) (+14.40%)   153162 ( 477.6) (-48.09%)
   test_operator_overloading:    2.877 (0.030)   334324 ( 135.7)    3.062 (0.044) ( +6.43%)   172689 ( 234.3) (-48.35%)
               test_pickling:    1.853 (0.030)   258136 ( 108.8)    2.116 (0.030) (+14.20%)   137702 ( 245.7) (-46.66%)
                test_pytypes:    2.088 (0.029)   270000 (  46.6)    2.725 (0.029) (+30.54%)   143312 ( 259.7) (-46.92%)
test_sequences_and_iterators:    4.152 (0.037)   377033 (  62.2)    4.408 (0.052) ( +6.15%)   216225 ( 211.2) (-42.65%)
              test_smart_ptr:    4.205 (0.036)   378643 (  69.6)    4.387 (0.042) ( +4.33%)   209329 ( 205.3) (-44.72%)
                    test_stl:    3.546 (0.032)   366140 (  89.6)    3.873 (0.033) ( +9.24%)   202854 ( 204.3) (-44.60%)
            test_stl_binders:   11.039 (0.090)   806419 ( 145.5)   11.174 (0.082) ( +1.22%)   439104 ( 215.3) (-45.55%)
      test_virtual_functions:    4.029 (0.032)   377739 (  74.1)    4.214 (0.045) ( +4.58%)   211784 ( 263.2) (-43.93%)
```

So gcc compiles everything faster, but uses almost twice as much memory when doing so.  (Linking is actually considerably faster in *wallclock* seconds under clang's lld as it runs in parallel, while regular `ld` does not).

jagerman commented:
This would likely also provide a way to deal with #948 (binding a class that doesn't support placement new).

jagerman commented on include/pybind11/factory_support.h:
```diff
@@ -0,0 +1,280 @@
+
+// Makes sure the `value` for the given value_and_holder is not allocated; if it is, deallocate it.
+// Returns the (null) value pointer reference.
+inline void *&deallocate(value_and_holder &v_h) {
```
One possible case is when you have overloaded constructors when you have mixed placement new and init factory constructors, such as:

```
cl.def("__init__", [](Class &self, double i) { new (&self) Class(i); });
cl.def("__init__", py::init([](int j) { return new Class(j); });
```
when called from Python as `Class(3)` to go to the second overload.  Because the lazy allocation happens during argument loading, the `&self` will get preallocated even though the overload fails, and the other init factories need to deal with it.

I suppose we could deprecated placement new constructors and get rid of this eventually (i.e. in an eventual 3.0), in which case we wouldn't need to worry about this case.

Re: comments, that's a fair criticism; I'll edit them to be more "why" than "what".

dean0x7d replied:
Ah, that makes sense with the mixed constructors. Deprecating the placement new constructors sounds good to me.

jagerman commented on include/pybind11/factory_support.h:
```diff
@@ -0,0 +1,280 @@
+        no_nullptr(ptr);
+        if (Py_TYPE(v_h.inst) != cl_type) {
+            // Inherited from on the Python side: an alias instance is required
+            if (!is_alias<Class>(ptr)) {
```
It should have been, but may have been dropped accidentally somewhere along the line.  I'll investigate.

jagerman commented on include/pybind11/factory_support.h:
```diff
@@ -0,0 +1,280 @@
+    static void construct(value_and_holder &v_h, Cpp<Class> &&result, PyTypeObject *) {
+        static_assert(std::is_move_constructible<Cpp<Class>>::value,
+            "pybind11::init() return-by-value factory function requires a movable class");
+        new (preallocate(v_h)) Cpp<Class>(std::move(result));
```
`preallocate` doesn't do anything if the value is already preallocated, so if you mix old-style placement-new `__init__`s with factory inits you get fewer allocation/deallocations.  If placement-new constructors are rare/deprecated, it seems a reasonable approach though.

(And on a side note, non-factory inits have a pretty big issue with the constructor being invoked multiple time if being used under multiple inheritance, so discouraging them is probably a good thing).

jagerman commented on include/pybind11/pybind11.h:
```diff
@@ -203,6 +199,8 @@ class cpp_function : public function {
                 a.descr = strdup(a.value.attr("__repr__")().cast<std::string>().c_str());
         }
 
+        rec->is_constructor = !strcmp(rec->name, "__init__") || !strcmp(rec->name, "__setstate__");
```
It was part of an earlier commit that was since dropped to fix the constructor signature when the first argument was a `handle`, but that got replaced with `v_h_proxy<Cpp<Class>>`.  I think you want me to go back to that, though.

jagerman replied:
The constructor signature rewriting got reintroduced with removal of the proxy class, so I kept this moved.

jagerman commented:
I've been a bit swamped with other things for the last few days, but I'll try to get this PR updated tomorrow.

jagerman commented:
The last commit just pushed should address the outstanding comments.

There is still some growth in .so size (~19KB) from the `py::init`-via-factory implementation.  This is buying us a fairly important fix for Python-side multiple inheritance (where multiple `__init__`s will invoke multiple C++ constructions on the same object) that can't be easily avoided with the previous placement-new implementation, so I think it's worthwhile.  I tried to extract some other bits into common code, but couldn't get anything that made a noticeable difference in `.so` size.

If all looks good, there's one last cleanup I think we should do which is to rename `factory_support.h` to `init_support.h` and also move the `struct init` and `struct init_alias` into it from `pybind11.h`.  Neither is very large, but they seem logically well matched with the factory code (and takes a bit of `detail` code out of the main header).

jagerman commented:
I've renamed this to `detail/init.h` to match #1001.  I also moved the various static functions in the factory class into free templated functions in the namespace, then used them to implement a more direct path for the common standard `py::init<...>()` constructors, which further cuts down the .so size increase (and should also make regular constructors slightly faster--they save the extra function pointer dereference and call that using factory functions would incur).

jagerman commented:
(The change looks bigger than it actually is; most of the commit diff is just from moving the static functions in the factory class out, and moving the `detail::init`/`detail::init_alias` structs into `detail/init.h`).

dean0x7d commented on include/pybind11/detail/init.h:
```diff
@@ -0,0 +1,333 @@
+// Fallback of the above for aliases without an `Alias(Cpp &&) constructor; does nothing and
+// returns false to signal the non-availability of the special constructor.
+template <typename Class, enable_if_t<!std::is_constructible<Alias<Class>, Cpp<Class> &&>::value, int> = 0>
+static constexpr bool construct_alias_from_cpp(value_and_holder &, Cpp<Class> &&) { return false; }
```
The control flow built around this feels a bit more complicated than it should be. This `return false` results in two different `throw type_error` messages below, but they have the same cause. Consider switching to `void` functions and place the `throw` statement in the `false` fallback. (I'm also using simple tag dispatch in the example below because it's a nice lightweight option, especially for binary overloads, and it also nicely maps to the future `if constexpr`.)

```c++
template <typename Class>
using is_alias_constructible = std::is_constructible<Alias<Class>, Cpp<Class> &&>;

template <typename Class>
void construct_alias_from_cpp(std::true_type /*is_alias_constructible*/,
                              value_and_holder &v_h, Cpp<Class> &&base) {
    deallocate(v_h) = new Alias<Class>(std::move(base));
}

template <typename Class>
[[noreturn]] void construct_alias_from_cpp(std::false_type /*is_alias_constructible*/,
                                           value_and_holder &, Cpp<Class> &&) {
    throw type_error("pybind11::init(): unable to convert returned instance to required "
                     "alias class: no `Alias<Class>(Class &&)` constructor available");
}
```

```c++
if (Class::has_alias && need_alias && !is_alias<Class>(ptr)) {
    ...
    construct_alias_from_cpp<Class>(is_alias_constructible<Class>{}, v_h, std::move(*ptr));
} else {
    deallocate(v_h) = ptr;
}
```

jagerman commented on tests/test_factory_constructors.py:
```diff
@@ -0,0 +1,451 @@
+                a = m.TestFactory2(tag.pointer, 1)
+                m.TestFactory1.__init__(a, tag.pointer)
+            elif bad == 2:
+                m.TestFactory1.__init__(NotPybindDerived.__new__(NotPybindDerived), tag.pointer)
```
I was looking for something allocated but uninitialized.  Will it hurt anything?

jagerman replied:
Updated to just pass an ordinary non-pybind instance (essentially matching to previous case, except for the non-pybind aspect).

dean0x7d replied:
It doesn't hurt anything I just thought it looked like unusual syntax for the functionally equivalent `NotPybindDerived()`. In general, for a native Python `class A`, the constructor `A() = A.__new__() + A.__init__()`. If `A` doesn't have `__init__()` then `A() == A.__new__()`, which is the case for `NotPybindDerived` in the test.

jagerman commented on tests/test_factory_constructors.py:
```diff
@@ -0,0 +1,451 @@
+
+@pytest.unsupported_on_py2
+@pytest.mark.parametrize('which, bad', [(1, 1), (1, 2), (6, 1), (6, 2), (6, 3), (6, 4)])
+def test_invalid_self(which, bad, capture):
```
That was to suppress the capture leak (which it does on my system, but apparently doesn't on one of the travis-ci builds).

I don't quite understand why that capture leak is happening at all, though.  Could the `mark.parametrized` be involved somehow?

jagerman replied:
Deparameterizing it worked around the issue.

jagerman commented:
Fixed up everything and squashed the commits down; I think this is ready to go (once the builds finish).

jagerman commented:
Merged!

wjakob commented on include/pybind11/detail/init.h:
```diff
@@ -0,0 +1,330 @@
+        auto *cl_type = get_type_info(typeid(Cpp<Class>));
+        cl.def("__init__", [cl_type](handle self_, Args... args) {
+            auto v_h = load_v_h(self_, cl_type);
+            if (v_h.instance_registered()) return; // Ignore duplicate __init__ calls (see above)
```
I stumbled across this comment (duplicate __init__ calls) while looking at ``init.h`` and was a bit puzzled. I wasn't able to find what the "see above" comment refers to. Would you mind clarifying this?

wjakob replied:
^ @jagerman

jagerman replied:
Ah, the code got shuffled around; the comment it's referring to is at lines 258/259 now.  I'll fix it up on `master`.

jagerman replied:
(They used to be right next to each other).

jagerman replied:
Fixed in 9f6a636e.

dean0x7d replied:
Ah, just got a conflict with those changes :) Those checks are gone now in #1014.

jagerman replied:
That was pretty incredibly bad timing :)
